src/main/java/com/puppycrawl/tools/checkstyle/CheckCloneService.java

```diff
@@ -0,0 +1,156 @@
+                        CHECK_MUST_IMPLEMENT_INTERFACE_MSG, checkClassName);
+                LOG.debug(message);
+            }
+            clone = check;
+        }
+
+        return clone;
+    }
+
+    /**
+     * Clones the given fileset check.
+     * @param originalCheck A fileset check to clone.
+     * @return Cloned fileset check.
+     * @noinspection InstanceofChain
+     */
+    // we cannot move this logic to the FileSetCheck subclasses,
+    // because they should not know about PackageObjectFactory
+    // and other configuration details
+    public static FileSetCheck cloneFileSetCheck(FileSetCheck originalCheck) {
+        final FileSetCheck clone;
+
+        if (originalCheck instanceof OneCheckInstancePerApplication) {
+            clone = originalCheck;
+        }
+        else if (originalCheck instanceof OneCheckInstancePerThread) {
+            final ClassLoader classLoader = CheckCloneService.class.getClassLoader();
+            final ModuleFactory factory = new PackageObjectFactory(
+                    Checker.class.getPackage().getName(), classLoader);
+            clone = doCloneCheck(originalCheck, factory);
+            clone.init();
+            clone.finishCloning(originalCheck);
+        }
+        else {
+            if (LOG.isDebugEnabled()) {
+                final String checkClassName = originalCheck.getClass().getSimpleName();
+                final String message = String.format(Locale.ROOT,
+                        CHECK_MUST_IMPLEMENT_INTERFACE_MSG, checkClassName);
+                LOG.debug(message);
+            }
+            clone = originalCheck;
+        }
+
+        return clone;
+    }
+
+    /**
+     * Clones a list of fileset checks.
+     * @param fileSetChecks A list of fileset checks to be cloned.
+     * @return A list of cloned fileset checks.
+     */
+    public static List<FileSetCheck> cloneFileSetChecks(List<FileSetCheck> fileSetChecks) {
+        return fileSetChecks.stream().map(CheckCloneService::cloneFileSetCheck)
+                    .collect(Collectors.toList());
+    }
+
+    /**
+     * Clones the check using the given module factory.
+     * @param check A check to clone.
+     * @param factory A factory for creating new checks.
+     * @param <T> The generic check type.
+     * @return The check clone.
+     */
+    private static <T extends Configurable & Contextualizable> T doCloneCheck(
+            T check, ModuleFactory factory) {
+        final Configuration config = check.getConfiguration();
+        final String moduleName = config.getName();
+        try {
+            final Object module = factory.createModule(moduleName);
+            final T clone = (T) module;
+            clone.contextualize(check.getContext());
+            clone.configure(config);
+            return clone;
+        }
+        catch (CheckstyleException ex) {
+            throw new IllegalStateException(
+                    "An unexpected exception raised while cloning check", ex);
```

Comment:
@rnveach Moved to the property file. Not sure if I could translate the message to other languages. Should I just translate them via some translation service or there is a translation group which could help with this task?